backend/src/user/user.service.ts

```diff
@@ -76,4 +76,33 @@ const updateCode = async (
 	return data[0]
 }
 
-export { getAllCodes, addCode, deleteCode, updateCode }
+const updateCodename = async (uid: string, codename: string, name: string) => {
+	if (codename.length > NAME_LEN_LIMIT)
+		throw new LargePayload('Code name exceed limit')
+	const transaction = await sequelize.transaction()
+	try {
+		const exist = await Code.findOne({
+			where: { uid, codename: name },
+			transaction,
+		})
+		if (exist) throw new ResourceConflict('Code name already exists')
+		const [affectedCount, data] = await Code.update(
+			{ codename: name },
+			{
+				where: {
+					uid,
+					codename,
+				},
+				returning: true,
+				transaction,
+			}
+		)
+		if (affectedCount == 0) throw new NotFound()
+		transaction.commit()
```

Comment:
Could there be a race condition here? E.g. if there are requests A and B, both to change "a.py" and "b.py" to "c.py".

A -> findOne doesn't find "c.py"
B -> findOne doesn't find "c.py"
A -> update "a.py" to "c.py"
B -> update "b.py" tp "c.py" (leads to conflict)

Reply:
By using transactions we can control the isolation level (which is currently set to the default `READ_UNCOMMITTED`) to a higher isolation level to lock the db from these race conditions.
However, currently I do not think this is necessary since if transaction B encounters a conflict it will just rollback the transaction as an error will be raised with duplicate composite key.
Since the transaction only have one write query there is not much cost to rollback and an error will be raised to the user. Also, as this case is highly unlikely (the same user opening multiple tabs?) I do not think it's worth the cost of db locking that would affect multiple users.